### DISCUSSION TITLE
AutoPiD with ZOE ZE50 ( Phase 2 )

### DISCUSSION THREAD

**Discussion by McFly98** [meatpiHQ/wican-fw#238] [Q&A] [answered]
Hello everyone,

I've been playing around with the wican for a while now. It's fun to learn the system layer by layer.

But now I'm not making any progress with my project.
I would like to use the autopid feature via MQTT for a Renault Zoe ZE50. Unfortunately, the PIDs I found don't help. The CAN bus has been significantly changed since 2020.

Challenge accepted.

So how do I find out the correct PIDs? Is there more to configure (Init)?
For initial attempts, I used the wican in the ELM327 emulation to record the data traffic with Carscanner on mqtt.
If I let the app read the SOC, I get this trace:
![image](https://github.com/user-attachments/assets/93629e10-885b-4234-aed0-0e1fe357064d)
What next steps can I take to expand the custom PIDs or the vehicle json?

Thanks for your help. A really great project here.

**Comment by TheSmartGerman** [comment]
May this helps: https://github.com/meatpiHQ/wican-fw/discussions/112

**Reply by McFly98** [reply]
Thanks for the hint.
But i already crawled through it.
The ZE50 seems to have other pids.

**Comment by TheSmartGerman** [comment, marked as answer]
Check OVMS: https://docs.openvehicles.com/en/latest/components/vehicle_renaultzoe_ph2_obd/docs/index.html

There's some read only support. Should be in the GitHub repository.

https://github.com/openvehicles/Open-Vehicle-Monitoring-System-3/tree/master/vehicle%2FOVMS.V3%2Fcomponents%2Fvehicle_renaultzoe_ph2_obd%2Fsrc

Check poller_ph2 file

**Reply by McFly98** [reply]
Hey thanks for your help.
I'am not sure about getting a pid from it but i will dig into.

**Reply by McFly98** [reply]
Thanks for the hints.

It runs now. But i don't understand where the "22" in the pid comes from. 
Here my custom pids:
![image](https://github.com/user-attachments/assets/41d6a2b6-3b15-437b-9193-16fa1711bba0)

I forked and changed the vehicle_profiles. For me the changes works.
I hope the pull-request is ok? 

Thanks for your help folks.

**Reply by meatpiHQ** [reply]
@McFly98 great work! you should create a new json file like I noted in the PR.

The first part of the PID request is the service, for EVs it's usually 22.

**Reply by McFly98** [reply]
Hello again, 

meanwhile i tried to dig deeper into.
I tried to get information about the climatecompressor.
OVMS has a definition for that:

`//HVAC
//{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIISESSION, SESSION_EXTDIAG,  { 0, 60, 60, 60 }, 0, ISOTP_STD }, // OBD Extended Diagnostic Session
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x4009,  { 0, 10, 60, 10 }, 0, ISOTP_STD }, // Cabin temp
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x4360,  { 0, 60, 60, 60 }, 0, ISOTP_STD }, // Cabin setpoint
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x43D8,  { 0, 10, 10, 5 }, 0, ISOTP_STD },  // Compressor speed
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x4402,  { 0, 10, 10, 10 }, 0, ISOTP_STD }, // Compressor state
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x4369,  { 0, 10, 10, 10 }, 0, ISOTP_STD }, // Compressor pressure in BAR
{ 0x744, 0x764, VEHICLE_POLL_TYPE_OBDIIEXTENDED, 0x4436,  { 0, 10, 10, 10 }, 0, ISOTP_STD }, // Compressor power`

Unfortunately Wican gets an timeout with the pid "224402".
I think it comes from ISOTP_STD. The other values are ISOTP_EXTFRAME. How can i set this parameter with wican?

Btw.: i figured out where the 22 comes from.
In the can [spezification](https://mk4-wiki.denkdose.de/artikel/can-bus/protokolle/uds/sid) the 0x22 command ist for reading Data by identifier.

**Reply by TheSmartGerman** [reply]
I guess u have to use individual pid init to change address..i see 0x744/0x764. I guess it's another ECU. Therefore another address i needed.

**Reply by meatpiHQ** [reply]
@McFly98 

As @TheSmartGerman mentioned you should use individual pid init. If you want to learn more about what are these command strings you should look at the ELM327 AT commands. I'm trying to add more documentation on this but it's not ready yet. Have a look at the links below.

https://www.elmelectronics.com/DSheets/ELM327DSH.pdf
https://meatpihq.github.io/wican-fw/config/automate/new_vehicle_profiles

**Reply by McFly98** [reply]
Thanks for your help.

I think i figured out the problem.
The most values are ATSP7 but some (the short adresses) are ATSP6.
If i only use the SP6 or SP7 values. Everything works fine, Even in different ECUs.
But if i use for example the SP7 values in the json and add SP6 values via custom init everything is messed up.
![Screenshot 2024-10-27 221812](https://github.com/user-attachments/assets/04e52515-46e9-4fc6-af2a-20ca54e4feb9)

![Screenshot 2024-10-27 221646](https://github.com/user-attachments/assets/2854b089-6cfa-49a8-b893-606c1734598e)
The same if i take both protocols in one json.
Is something more needed to switch the Protocol in one configuration?

**Reply by meatpiHQ** [reply]
@McFly98 So for switching protocols you'll need these commands:

```
Set protocol: ATSP
Set Header for Flow control: ATSHFC
CAN Receive Address: ATCRA
Set Header for Flow Control: ATFCSH
Flow Control Set Mode: ATFCSM
```
You can find tests commands in car scanner logs.

**Reply by McFly98** [reply]
Thanks.

but i didn't find the ATSHFC command.
After some hours of trying i figured out a config that works.
But only via the cutoms pids. 
For the moment i will let it this way:
![Screenshot 2024-10-28 222544](https://github.com/user-attachments/assets/2205e460-4c14-475c-ab19-61028df39335)

**Reply by meatpiHQ** [reply]
@McFly98 Can you please share the init strings for each PID?

**Reply by McFly98** [reply]
Tada ;)


<details><summary>Zoe Custom PIDs</summary>
<p>

```python
{
  "initialisation": "ATZ;ATE0;ATH1;ATS0;ATM0;ATAT1;ATST26;ATCP18;ATFCSD300000;",
  "grouping": "disable",
  "destination": "",
  "cycle": "5000",
  "car_specific": "disable",
  "ha_discovery": "disable",
  "car_model": "",
  "pids": [
    {
      "Name": "Odometer",
      "Init": "ATSP7;ATFCSH18DADAF1;ATCRA18DAF1DA;ATSHDADAF1;ATFCSM1;",
      "PID": "222006",
      "Expression": "[B4:B6]",
      "Period": "10000",
      "Send_to": "wican/zoe/Odometer",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "Plugged",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "222B85",
      "Expression": "B4",
      "Period": "2500",
      "Send_to": "wican/zoe/Plugged",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "Soc",
      "Init": "ATSP7;ATSHDADBF1;ATFCSH18DADBF1;ATFCSM1;ATCRA18DAF1DB;",
      "PID": "229002",
      "Expression": "[B4:B5] * 0.01",
      "Period": "5000",
      "Send_to": "wican/zoe/Soc",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "PowerAvail",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "22300F",
      "Expression": "[B4:B5] * 0.025",
      "Period": "5000",
      "Send_to": "wican/zoe/PowerAvail",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "AcPhases",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "22300B",
      "Expression": "B4 + 1",
      "Period": "5000",
      "Send_to": "wican/zoe/AcPhases",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "12vBat",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "222005",
      "Expression": "[B4:B5] * 0.01",
      "Period": "5000",
      "Send_to": "wican/zoe/12vBat",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "AcCurrent",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "22300D",
      "Expression": "[B4:B5] * 0.1",
      "Period": "2500",
      "Send_to": "wican/zoe/AcCurrent",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "AcVoltage",
      "Init": "ATSP7;ATSHDADAF1;ATFCSH18DADAF1;ATFCSM1;ATCRA18DAF1DA;",
      "PID": "222B8A",
      "Expression": "[B4:B5] * 0.5",
      "Period": "2500",
      "Send_to": "wican/zoe/AcVoltage",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "Soh",
      "Init": "ATSP7;ATSHDADBF1;ATFCSH18DADBF1;ATFCSM1;ATCRA18DAF1DB;",
      "PID": "229003",
      "Expression": "[B4:B5] * 0.01",
      "Period": "10000",
      "Send_to": "wican/zoe/Soh",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "CompState",
      "Init": "ATSP6;ATFCSH744;ATCRA764;ATSH744;ATFCSM1;",
      "PID": "224402",
      "Expression": "B4",
      "Period": "5000",
      "Send_to": "wican/zoe/CompState",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "Locked",
      "Init": "ATSP6;ATFCSH745;ATCRA765;ATSH745;ATFCSM1;",
      "PID": "228004",
      "Expression": "B4",
      "Period": "10000",
      "Send_to": "wican/zoe/Locked",
      "Type": "MQTT_Topic"
    },
    {
      "Name": "CabinTemp",
      "Init": "ATSP6;ATFCSH744;ATCRA764;ATSH744;ATFCSM1;",
      "PID": "224009",
      "Expression": "([B4:B5] - 400) / 10",
      "Period": "10000",
      "Send_to": "wican/zoe/CabinTemp",
      "Type": "MQTT_Topic"
    }
  ]
}
``` 

</p>
</details>
